Theming description for ownCloud Web (#53)
owncloud/docs-webui: modules/owncloud_web/pages/themed_webui.adoc

= ownCloud Web with Custom Theming
:toc: right

> :toc-levels: 1

Reviewer: ```suggestion
:toclevels: 1
```

== Introduction

{description}

This can be achieved by providing a json file that contains text snippets (like brand name and slogan), paths to images (e.g. logos or favicon) and design tokens for various parameters like color, sizing and spacing. In the following we explain the setup and configuration options and provide an empty template for you to get started.

== Providing a Theme

Your theming configuration resides in a `.json` file, e.g. `theme.json`. To load this file, it needs to be correctly referenced inside your `config/config.json`. An example configurations can be found on {example-theme-url}[GitHub].

NOTE: If no theme is provided or the loading of your custom theme fails, the standard ownCloud theme will be loaded as a fallback. However, this doesn’t stop you from correctly loading a theme that is wrongly formatted, so please read the instructions below carefully.

To reference your theme, you have two options:

> * Using a URL, e.g.: "theme": "https://externalurl.example.com/theme-name/theme.json",. To avoid CORS issues, make sure that you host the URL on the same domain as your ownCloud web hosting.

Reviewer: ```suggestion
* Using a URL like:
+
[source,plaintext]
----
"theme": \"https://externalurl.example.com/theme-name/theme.json",
----
+
To avoid CORS issues, make sure that you host the URL on the same domain as your ownCloud web hosting.
```
Making this a source block plus make the link not render as link

> * For development and testing purposes, you can store your `theme.json` inside `packages/web-runtime/themes/{theme-name}/` and reference it in the `config.json` file. However, this isn’t recommended for production use since your changes may get lost when updating Infinite Scale or the Web app on ownCloud 10.x.

Reviewer: ```suggestion
* For development and testing purposes, you can store your `theme.json` inside \`packages/web-runtime/themes/{theme-name}/` and reference it in the `config.json` file. However, this isn’t recommended for production use since your changes may get lost when updating Infinite Scale or the Web app on ownCloud 10.x.
```
Needs escaping, else the attribute will get rendered, is empty and will return a build warning

== Configuring a Theme

Inside your `theme.json` file, you can provide multiple themes as first-level objects. Currently, only the one called "default" gets applied when the frontend application is started.

Feel free to use the snippet below as a base for writing your own theme. Replace the strings and image file paths accordingly. Also, make sure to delete the comments from the file.

[source,plaintext]
----
{
  "default": {
    "general": {
      "name": "ownCloud",
      "slogan": "ownCloud – A safe home for all your data"
    },
    "logo": {
      "topbar": "https://externalurl.example.com/url/for/remote/theme/assets/logo.svg",
      "favicon": "https://externalurl.example.com/url/for/remote/theme/assets/favicon.jpg",
      "login": "relative/path/for/local/theme/logo.svg"
    },
    "loginPage": {
      "autoRedirect": true,
      "backgroundImg": "relative/path/for/local/theme/background.jpg"
    },
    "designTokens": {}
  },
  "alternative": {},
  "dark": {}
}
----

See below for the meaning of all the first-level objects inside a single theme and recommendations on how to make best use of them.

=== "general"

> Here, you can specify a `"name"` and a `"slogan"` string. The name gets used in the HTML page `<title>`, and both of them are shown on various screens (e.g. login, loading, error and public share pages).

Reviewer: ```suggestion
Here, you can specify a `"name"` and a `"slogan"` string. The name gets used in the HTML page `<title>`, and both of them are shown on various screens (like: login, loading, error and public share pages).
```

=== "designTokens"

To further customize your ownCloud instance, you can provide your own styles. To give you an idea of how a working design system looks, feel free to head over to the {design-tokens-url}[ownCloud design tokens] for inspiration.

TIP: All the variables are initialized using the ownCloud design tokens and then overwritten by your theme variables. Therefore, you don’t have to provide all the variables and can use the default ownCloud colors as a fallback.

In general, the theme loader looks for a `designTokens` key inside your theme configuration. Inside the `designTokens`, it expects to find a collection of `colorPalette`, `fontSizes` and `spacing`. The structure is outlined below:

[source,plaintext]
----
{
  "default": {
    "general": {},
    "designTokens": {
      "breakpoints": {},
      "colorPalette": {},
      "fontSizes": {},
      "sizes": {},
      "spacing": {}
    }
  }
}
----

Follow this structure to make sure your theming configuration is loaded correctly.

==== Extendability

> If you define different key-value pairs inside any of the objects in `"designTokens"`, they will get loaded and initialized as CSS custom properties but don’t take any effect in the user interface. This gives you an opportunity to customize extensions from within the theme in the web runtime (and not the extension itself).

Reviewer: Needs clarification what is meant with extension

    "text-inverse": "",
    "text-muted": ""
  }
}
----

==== Font Sizes

You can change the `default`, `large` and `medium` font sizes according to your needs. If you need more customization options regarding font sizes, open an issue on {gh-webui-url}[GitHub] with a detailed description.

Font size variables get prepended with `--oc-font-size-`, so e.g. `“default”` creates the custom CSS property `--oc-font-size-default`.

[source,plaintext]
----
{
  "fontSizes": {
    "default": "",
    "large": "",
    "medium": ""
  }
}
----

==== Sizes

Use sizing variables to change various UI elements, such as icon and logo appearance, table row or checkbox sizes, according to your needs. If you need more customization options regarding sizes, open an issue on {gh-webui-url}[GitHub] a detailed description.

Size variables get prepended with `--oc-size-`, so e.g. `“icon-default”` creates the custom CSS property `--oc-size-icon-default`.

[source,plaintext]
----
{
  "sizes": {
    "form-check-default": "",
    "height-small": "",
    "height-table-row": "",
    "icon-default": "",
    "max-height-logo": "",
    "max-width-logo": "",
    "width-medium": ""
  }
}
----

==== Spacing

> Use the six spacing options (`xsmall | small | medium | large | xlarge | xxlarge`) to create a more (or less) condensed version of the user interface. If you need more customization options regarding sizes, open an issue on {gh-webui-url}[GitHub] with a detailed description.

Reviewer 2: ```suggestion
Use the six spacing options (`xsmall | small | medium | large | xlarge | xxlarge`) to create a more (or less) condensed version of the user interface. If you need more customization options regarding spacing, open an issue on {gh-webui-url}[GitHub] with a detailed description.
```
